[Dependency Analysis] Remove Unused Dependencies

ParaskP7 opened the pull request:
Project Thread: paaHJt-6YV-p2
Depends On:
- [x] [ADC#35](https://github.com/Automattic/android-dependency-catalog/pull/35)
- [x] Update ADC to `1.21.0`

### Description

Based on the `unused` related advises generated by the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin, this PR removes all unused dependencies from this project.

FYI: As part of this change, the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin got also updated to its latest [1.33.0](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CHANGELOG.md#version-1330) version.

### Testing Steps

1. Tooling:
    - Run the `./gradlew buildHealth` task and verify that there is no `unused` related advise remaining within both reports, JSON and txt included:
        - `build-health-report.json` -> Search for `unusedCount`
        - `build-health-report.txt` -> Search for `unused`
    - Check the manually triggered [scheduled build](https://buildkite.com/automattic/wordpress-fluxc-android/builds/3804) and verify that everything works as expected with this newer version of the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin ([1.33.0](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CHANGELOG.md#version-1330)).
3. Testing:
    - Smoke test the `FluxC Example` app, try out all available screens and functionality. Verify everything is working as expected.
    - Also, if you want to be thorough about reviewing the changes, you could quickly smoke test, either the [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android) and/or [WCAndroid](https://github.com/woocommerce/woocommerce-android), with this version of `FluxC`, or via composite builds, and see if it works as expected.

ParaskP7 committed:
Deps: Update dependency analysis to 1.33.0 (via catalog's pr-sha comb)

FYI: This change was done for testing purposes and until the below
'Android Dependency Catalog' #35 PR gets merged to 'trunk'. When that's
done, the 'catalogVersion' will be updated to '1.21.0' instead.

Android Dependency Catalog PR: [Dependency Updates] Update
'dependency-analysis' to 1.33.0 #35
- https://github.com/Automattic/android-dependency-catalog/pull/35

ParaskP7 committed:
Deps: Remove unused mockito android on android test from example

It is unclear when and why this 'androidTest' related dependency got
added to this project, as one can only navigate to the initial commit
(78b37f9856bd3b2ed7a30f420885c22afab13cd0) and still see that there.
Mockito imports are only to be found within the various 'test' related
source sets and never within the 'androidTest' related source sets.

Also, these two 'System.setProperty("dexmaker.dexcache", ...)' lines,
which were needed for Mockito, are also removed as well for the same
reason.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :example
Unused dependencies which should be removed:
  androidTestImplementation sharedLibs.mockito.android
  ...

ParaskP7 committed:
Deps: Remove unused google dagger compiler on android test from example

Since the same dependency is already defined on the 'kapt' level,
defining it again on the 'kaptAndroidTest' level is redundant.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :example
Unused dependencies which should be removed:
  ...
  kaptAndroidTest sharedLibs.google.dagger.compiler

ParaskP7 committed:
Deps: Update androidx room dependencies in example

------------------------------------------------------------------------

This update was suggested by the dependency analysis report, see below:

Advice for :example
Unused dependencies which should be removed:
  ...
  implementation sharedLibs.androidx.room.ktx

These transitive dependencies should be declared directly:
  ...
  androidTestImplementation sharedLibs.androidx.room.runtime
  ...
  testImplementation sharedLibs.androidx.room.runtime

Dependencies which should be removed or changed to runtime-only:
  ...
  runtimeOnly sharedLibs.androidx.room.runtime (was implementation)
  ...

Unused annotation processors that should be removed:
  kapt sharedLibs.androidx.room.compiler

ParaskP7 committed:
Deps: Replace androidx legacy support v4 dependency with sfl in example

Removing the 'androidx.legacy.support.v4' dependency wasn't enough by
itself as the build was then failing. Adding the transient
'androidx.swiperefreshlayout:swiperefreshlayout' dependency was actually
required in order to make the build successful again.

FYI: This change is also updating the 'catalogVersion' to that catalog's
pr-sha combination where the 'androidx.swiperefreshlayout' dependency is
now available.

------------------------------------------------------------------------

This replacement was suggested by the dependency analysis report,
see below:

Advice for :example
Unused dependencies which should be removed:
  ...
  implementation sharedLibs.androidx.legacy.support.v4

These transitive dependencies should be declared directly:
  ...
  implementation 'androidx.swiperefreshlayout:swiperefreshlayout:1.1.0'

ParaskP7 committed:
Deps: Remove unused androidx arch core dependencies from example

The only 'androidx.arch.core' related dependency that is needed, and
only on the 'testImplementation' and 'androidTestImplementation' level
is the 'androidx.arch.core.testing' for testing purposes and in order to
utilize the 'InstantTaskExecutorRule' rule.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :example
Unused dependencies which should be removed:
  implementation sharedLibs.androidx.arch.core.common
  implementation sharedLibs.androidx.arch.core.runtime

ParaskP7 committed:
Deps: Replace eventbus dependency with eventbus java in fluxc

Removing the 'eventbus' dependency wasn't enough by itself as the build
was then failing. Adding the transient 'eventbus-java' dependency was
actually required in order to make the build successful again.

FYI: This change is also updating the 'catalogVersion' to that catalog's
pr-sha combination where the 'eventbus-java' dependency is now
available.

------------------------------------------------------------------------

This replacement was suggested by the dependency analysis report,
see below:

Advice for :fluxc
Unused dependencies which should be removed:
  api sharedLibs.eventbus
  ...

These transitive dependencies should be declared directly:
  ...
  api 'org.greenrobot:eventbus-java:3.3.1'
  ...

ParaskP7 committed:
Deps: Remove androidx appcompat dependency from fluxc

Only the 'example' module is using such UI related 'androidx'
dependencies.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :fluxc
Unused dependencies which should be removed:
  implementation sharedLibs.androidx.appcompat
  ...

ParaskP7 committed:
Deps: Remove androidx recyclerview dependency from fluxc

Only the 'example' module is using such UI related 'androidx'
dependencies.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :fluxc
Unused dependencies which should be removed:
  implementation sharedLibs.androidx.recyclerview
  ...

ParaskP7 committed:
Deps: Remove assertj core dependency on test from fluxc

Only the 'example' and 'plugins.woocommerce' modules are using 'assertj'
assertions.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :fluxc
Unused dependencies which should be removed:
  testImplementation sharedLibs.assertj.core

ParaskP7 committed:
Build: Update dependency analysis configuration on fluxc processor

All 'unused' related advises for the 'fluxc-processor' are incorrect
(for an unknown to me reason) and as such those were instead suppressed.
The 'onUnusedDependencies', 'onUnusedAnnotationProcessors' and
'onRedundantPlugins' exclusion configuration were used to achieve that.

------------------------------------------------------------------------

Advice for :fluxc-processor
Unused dependencies which should be removed:
  api 'org.jetbrains.kotlin:kotlin-stdlib:1.9.25'
  implementation sharedLibs.google.autoService

Unused annotation processors that should be removed:
  kapt sharedLibs.google.autoService

Unused plugins that can be removed:
  kotlin-kapt: this project has the kotlin-kapt
   (org.jetbrains.kotlin.kapt) plugin applied, but there are no used
   annotation processors.

wzieba commented on fluxc-processor/build.gradle:
```diff
@@ -20,6 +20,21 @@ dependencies {
+dependencyAnalysis {
+    issues {
+        onUnusedDependencies {
+            exclude("org.jetbrains.kotlin:kotlin-stdlib")
```
It's true that stdlib is not used in this module: can we indeed exclude adding this by Kotlin Gradle plugin by introducing `gradle.properties` file with `kotlin.stdlib.default.dependency=false` ?

ParaskP7 replied:
Interesting point @wzieba thanks! 👍 

Thoughts:
1. Hmmm, do you think it will work, have you tried it? 🤔 
2. I wonder if it is better to just keep it as is, and exclude it on the Dependency Analysis level, than changing `gradle.properties` for the whole project for that matter, just because of `fluxc-processor` and the fact that this module is legacy. Although this change doesn't seem risky, I think it is better to leave this alone, not sure... 🤔

If you really think this is a good move, I'll go ahead and make this change anyway. 👍

ParaskP7 replied:
UPDATE: As discussed during our Android Core catch-up, I am going to try and replace Kotlin `kapt` with Java `annotationProcessor` and completely remove the unused `s3` plugins from this and the `fluxc-annotations` modules.

wzieba replied:
Thanks @ParaskP7 ! My one request would be to verify if performance wasn't affected much (maybe gradle-profiler with change of a code that requires `fluxc-annotations` to regenerate code?)

ParaskP7 replied:
That's a good idea @wzieba , will try gradle-profiler and see how it goes, wise me luck! 🍀

ParaskP7 replied:
Okay @wzieba , this is now done!

1. 5893fa42d560d5f08beb7f221bd81085c220d969

I couldn't benchmark incremental build as there was no `apply-x-change-to` for pure `.txt` file (ie. targeting `wc-wp-api-endpoints.txt`) that Gradle Profiler can provide us, nor a custom scenario for that matter. The closer we could get is `apply-property-resource-change-to` but this adds/removes property entries (ie. `xyz=123`). 🤔 

Also, I couldn't think of any other better place to update the code and test an incremental build as we don't want to verify compilation, but rather code generation. 🤷 

However, I did benchmark at least normal builds (via `fluxc-processor.scenarios`) and everything seems to be okay, see diff below:

<details>
    <summary>fluxc-processor.scenarios</summary>

```txt
fluxc-processor-with-kapt {
    title = "FluxC Processor with KAPT"
    tasks = ["assembleDebug"]
    gradle-args = ["--rerun-tasks"]
    # Caused by: java.lang.IllegalArgumentException: Can only modify Java or Kotlin source files
    # apply-abi-change-to = ["fluxc-processor/src/main/resources/wc-wp-api-endpoints.txt"]
    # Caused by: java.lang.IllegalArgumentException: not a valid name: org_acme_some=_5b4c6680_3d59_411a_9da0_4b684648a239_fluxc_processor_with_kapt_2ed5a0f4_WARM_UP_1
    # apply-property-resource-change-to = ["fluxc-processor/src/main/resources/wc-wp-api-endpoints.txt"]
}

fluxc-processor-with-annotation-processor {
    title = "FluxC Processor with Annotation Processor"
    tasks = ["assembleDebug"]
    gradle-args = ["--rerun-tasks"]
}
```

</details>

![image](https://github.com/user-attachments/assets/3881cfe7-ed2b-4daf-879c-2688b4b31c9d)


2. fc25de28672848a2d918c2f4f4cab8b6b39775b7 + f1e4b62e257b720f127d1b95f63d105fc42c3051

In addition to the above, the unnecessary `publishToS3` configuration got removed from both, the `fluxc-annotations` and `fluxc-processor` modules to simply the overall configuration, plus win some CI time back for us.

PS: Before merging (or reverting) this let's wait to get a response back from @oguzkocer (see interal thread: p1723556208848519-slack-C02KLTL3MKM).

ParaskP7 replied:
UPDATE: Having the build succeed with only publishing `fluxc` and `plugins:woocommerce`, I then tried using the `<PR>-<SHA>` combination on the WCAndroid project, this actually quickly failed with the below:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':WooCommerce:dataBindingMergeDependencyArtifactsJalapenoDebug'.
> Could not resolve all files for configuration ':WooCommerce:jalapenoDebugCompileClasspath'.
   > Could not find FluxC:fluxc-annotations:unspecified.
     Required by:
         project :WooCommerce > org.wordpress:fluxc:3075-f1e4b62e257b720f127d1b95f63d105fc42c3051
         project :WooCommerce > org.wordpress.fluxc.plugins:woocommerce:3075-f1e4b62e257b720f127d1b95f63d105fc42c3051
```

It seems that the published artifacts for `fluxc-annotations` and `fluxc-processor` are actually needed, I just don't understand enough about the `s3` configuration to derive the why...

ParaskP7 replied:
UPDATE: Actually, maybe this (ffa0b2c315cace54c8dd1d104ee35ba69e44c104) was the configuration I was missing to make the above working. 😅 Testing...

ParaskP7 replied:
UPDATE: Nope, the above didn't help and the WCAndroid build is still failing.

FYI: I also noticed that the `import org.wordpress.android.fluxc.annotations.action.Action` is the only `fluxc-annotations` specific import being used directly on the WCAndroid project, just like on the JP/WPAndroid project.

ParaskP7 replied:
FYI: Having discussing it with @oguzkocer and @wzieba I am now dropping those 3 commits that were removing the seemingly unnecessary publishing to `s3` of the `fluxc-annotations` and `fluxc-processors` artifacts.

PS: For more info see: p1723556208848519-slack-C02KLTL3MKM

ParaskP7 committed:
Deps: Update catalog to 1.21.0 (dependency analysis 1.33.0)

Catalog: https://github.com/Automattic/android-dependency-catalog/
releases/tag/1.21.0

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/remove-unused-dependencies

ParaskP7 committed:
Build: Replace kapt with annotation processor for fluxc processor

This is done in order to make this pure Java module ends-up becoming as
simple as possible and avoid tools like Dependency Analysis falsely
suggesting that there exist unused dependencies where there is none.

FYI: KAPT was added to this module via this
(e6feceab73347b2c07990ddfa98e0d627f5a43b3) commit, because of the fact
that the 'AutoService' annotation processor got split into two packages.
The assumption here is that they could have used 'annotationProcessor'
back then, that is, instead of 'kapt' and adding an unnecessary
dependency to the 'kotlin' plugin altogether.

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/remove-unused-dependencies
